Improve parameter substitution for database log

cproensa opened the pull request:
For showing the database query log, only mysql style parameters "?" were
supported. Other like postgres "$nnn" or oracle ":xxx" were not treated.

Change the token substitution to allow other token types to be replaced
with the parameter value for log output.

Fixes: #22005

dregad commented on core/database_api.php:
```diff
@@ -409,36 +409,46 @@ function db_query( $p_query, array $p_arr_parms = null, $p_limit = -1, $p_offset
 		if( !empty( $p_arr_parms ) ) {
-			while( preg_match( '/\?/', $p_query, $t_matches, PREG_OFFSET_CAPTURE, $t_lastoffset ) ) {
-				$t_matches = $t_matches[0];
+			# For mysql, tokens are '?', and parameters are binded sequentially
```
*bound* not binded...

dregad commented on core/database_api.php:
```diff
@@ -409,36 +409,46 @@ function db_query( $p_query, array $p_arr_parms = null, $p_limit = -1, $p_offset
+			# For mysql, tokens are '?', and parameters are binded sequentially
+			# For pgsql, tokens are '$number', and parameters are binded by the denoted index (1-based) in the parameter array
+			# For oracle, tokens are ':string', but ADOdb treat them as sequential in appearance order, so they behave like mysql
+			while( preg_match( '/(?<token>\?|\$|:)(?<index>[0-9]*)/', $p_query, $t_matches, PREG_OFFSET_CAPTURE, $t_lastoffset ) ) {
```
Wrap long lines to avoid horizontal scrolling.
For the while loop, maybe set `$t_regex '/(?<token>\?|\$|:)(?<index>[0-9]*)/';` 
then `while( preg_match( $t_regex, ... `

dregad commented on core/database_api.php:
```diff
@@ -409,36 +409,46 @@ function db_query( $p_query, array $p_arr_parms = null, $p_limit = -1, $p_offset
 				if( $i <= count( $p_arr_parms ) ) {
-					if( is_null( $p_arr_parms[$i] ) ) {
+					if( 'pgsql' == $t_db_type ) {
+						# For pgsql, the binded value is indexed by the parameter name
```
*bound*

cproensa commented:
@dregad thanks for the language tips.
Also, moved the logic outside of db_query(), to clean it up.

dregad commented on core/database_api.php:
```diff
@@ -1323,3 +1287,60 @@ function db_mysql_fix_utf8( $p_string ) {
+ * Process a query string by replacing token parameters by their bound values
+ * @param string $p_query		Query string
+ * @param array $p_arr_parms	Parameter array
+ * @return string				Processed query string
```
Not lining up - use spaces here instead of tabs

dregad commented on core/database_api.php:
```diff
@@ -1323,3 +1287,60 @@ function db_mysql_fix_utf8( $p_string ) {
+	if( !empty( $p_arr_parms ) ) {
+		# For mysql, tokens are '?', and parameters are bound sequentially
+		# For pgsql, tokens are '$number', and parameters are bound by the denoted index (1-based) in the parameter array
+		# For oracle, tokens are ':string', but ADOdb treat them as sequential in appearance order, so they behave like mysql
```
Comment lines are still too long, please wrap

dregad commented on core/database_api.php:
```diff
@@ -1323,3 +1287,60 @@ function db_mysql_fix_utf8( $p_string ) {
+					}
+				} else {
+					echo( 'Invalid argument type passed to query_bound(): ' . ( $i + 1 ) );
+					exit( 1 );
```
Not part of your change, but do we really need to *exit* here ? 
Maybe a warning is sufficient, what do you think ?

dregad commented on core/database_api.php:
```diff
@@ -1323,3 +1287,60 @@ function db_mysql_fix_utf8( $p_string ) {
+				}
+				$p_query = utf8_substr( $p_query, 0, $t_utf8_offset )
+						. $t_replace
+						. utf8_substr( $p_query, $t_utf8_offset + utf8_strlen( $t_match_param[0] ) );
```
Remove one indent tab

cproensa commented:
@dregad 
fixed the cosmetic issues
Regarding the "exit", i have removed it since it makes no sense to break execution at this point. If the query fails due to incorrect parameters, this is not the place to stop.
